flang/runtime/execute.cpp

```diff
@@ -0,0 +1,184 @@
+    const Descriptor *cmdmsg, const char *sourceFile, int line) {
+  Terminator terminator{sourceFile, line};
+  const char *newCmd{EnsureNullTerminated(
+      command.OffsetElement(), command.ElementBytes(), terminator)};
+
+  if (exitstat) {
+    RUNTIME_CHECK(terminator, IsValidIntDescriptor(exitstat));
+  }
+
+  if (cmdstat) {
+    RUNTIME_CHECK(terminator, IsValidIntDescriptor(cmdstat));
+    // Assigned 0 as specifed in standard, if error then overwrite
+    StoreIntToDescriptor(cmdstat, CMD_EXECUTED, terminator);
+  }
+
+  if (cmdmsg) {
+    RUNTIME_CHECK(terminator, IsValidCharDescriptor(cmdmsg));
+  }
+
+  if (wait) {
+    // either wait is not specified or wait is true: synchronous mode
+    int status{std::system(newCmd)};
+    int exitStatusVal{TerminationCheck(status, cmdstat, cmdmsg, terminator)};
+    // If sync, assigned processor-dependent exit status. Otherwise unchanged
+    CheckAndStoreIntToDescriptor(exitstat, exitStatusVal, terminator);
+  } else {
+// Asynchronous mode
+#ifdef _WIN32
+    STARTUPINFO si;
+    PROCESS_INFORMATION pi;
+    ZeroMemory(&si, sizeof(si));
+    si.cb = sizeof(si);
+    ZeroMemory(&pi, sizeof(pi));
+
+    // append "cmd.exe /c " to the beginning of command
+    const char *prefix{"cmd.exe /c "};
+    char *newCmdWin{(char *)AllocateMemoryOrCrash(
+        terminator, std::strlen(prefix) + std::strlen(newCmd) + 1)};
+    std::strcpy(newCmdWin, prefix);
+    std::strcat(newCmdWin, newCmd);
+
+    // Convert the char to wide char
+    const size_t sizeNeeded{mbstowcs(NULL, newCmdWin, 0) + 1};
+    wchar_t *wcmd{new wchar_t[sizeNeeded]};
+    if (std::mbstowcs(wcmd, newCmdWin, sizeNeeded) == static_cast<size_t>(-1)) {
+      terminator.Crash("Char to wide char failed for newCmd");
+    }
+    FreeMemory((void *)newCmdWin);
+
+    if (CreateProcess(nullptr, wcmd, nullptr, nullptr, FALSE, 0, nullptr,
+            nullptr, &si, &pi)) {
+      // Close handles so it will be removed when terminated
+      CloseHandle(pi.hProcess);
+      CloseHandle(pi.hThread);
+    } else {
+      if (!cmdstat) {
+        terminator.Crash(
+            "CreateProcess failed with error code: %lu.", GetLastError());
+      } else {
+        StoreIntToDescriptor(cmdstat, (uint32_t)GetLastError(), terminator);
+        CheckAndCopyCharToDescriptor(cmdmsg, "CreateProcess failed.");
+      }
+    }
+    FreeMemory((void *)wcmd);
+#else
+    // terminated children do not become zombies
+    signal(SIGCHLD, SIG_IGN);
+    pid_t pid{fork()};
+    if (pid < 0) {
+      if (!cmdstat) {
+        terminator.Crash("Fork failed with pid: %d.", pid);
+      } else {
+        StoreIntToDescriptor(cmdstat, FORK_ERR, terminator);
+        CheckAndCopyCharToDescriptor(cmdmsg, "Fork failed");
+      }
+    } else if (pid == 0) {
+      int status{std::system(newCmd)};
+      TerminationCheck(status, cmdstat, cmdmsg, terminator);
+      exit(status);
+    }
+#endif
+    FreeMemory((void *)newCmd);
```

Comment:
Move this deallocation out one level of braces so that it also handles the synchronous case.

Reply:
Done